protocol/skeleton.go

```diff
@@ -29,105 +30,67 @@ func getHeaders(clt proto.V1Client, req *proto.GetHeadersRequest) ([]*types.Head
 }
 
 type skeleton struct {
-	slots []*slot
-	span  int64
-	num   int64
-}
-
-func (s *skeleton) LastHeader() *types.Header {
-	slot := s.slots[len(s.slots)-1]
-
-	return slot.blocks[len(slot.blocks)-1].Header
-}
-
-func (s *skeleton) build(clt proto.V1Client, ancestor types.Hash) error {
-	// since ancestor is the common block we need to query the next one
-	headers, err := getHeaders(clt, &proto.GetHeadersRequest{Hash: ancestor.String(), Skip: s.span - 1, Amount: s.num})
-	if err != nil {
-		return err
-	}
-	s.addSkeleton(headers) // nolint
-
-	return nil
+	blocks []*types.Block
+	skip   int64
+	amount int64
 }
 
-func (s *skeleton) fillSlot(indx uint64, clt proto.V1Client) error {
-	slot := s.slots[indx]
-	req := &proto.GetHeadersRequest{
-		Hash:   slot.hash.String(),
-		Amount: s.span,
-	}
-	resp, err := getHeaders(clt, req)
-
+// getBlocksFromPeer fetches the blocks from the peer,
+// from the specified block number (including)
+func (s *skeleton) getBlocksFromPeer(
+	peerClient proto.V1Client,
+	initialBlockNum uint64,
+) error {
+	// Fetch the headers from the peer
+	headers, err := getHeaders(
+		peerClient,
+		&proto.GetHeadersRequest{
+			Number: int64(initialBlockNum),
+			Skip:   s.skip,
+			Amount: s.amount,
+		},
+	)
 	if err != nil {
 		return err
 	}
 
-	slot.blocks = []*types.Block{}
-
-	for _, h := range resp {
-		slot.blocks = append(slot.blocks, &types.Block{
-			Header: h,
-		})
-	}
-
-	// for each header with body we request it
-	bodyHashes := []types.Hash{}
-	bodyIndex := []int{}
-
-	for indx, h := range resp {
-		if h.TxRoot != types.EmptyRootHash {
-			bodyHashes = append(bodyHashes, h.Hash)
-			bodyIndex = append(bodyIndex, indx)
+	// Make sure the number sequences match up
+	for i := 1; i < len(headers); i++ {
+		if headers[i].Number-headers[i-1].Number != 1 {
+			return errors.New("invalid header sequence")
 		}
 	}
 
-	if len(bodyHashes) == 0 {
-		return nil
+	// Construct the body request
+	headerHashes := make([]types.Hash, len(headers))
+	for index, header := range headers {
+		headerHashes[index] = header.Hash
 	}
 
-	bodies, err := getBodies(context.Background(), clt, bodyHashes)
+	getBodiesContext, cancelFn := context.WithTimeout(
+		context.Background(),
+		time.Second*10,
+	)
+	defer cancelFn()
+
+	// Grab the block bodies
+	bodies, err := getBodies(getBodiesContext, peerClient, headerHashes)
 	if err != nil {
 		return err
 	}
 
-	for indx, body := range bodies {
-		slot.blocks[bodyIndex[indx]].Transactions = body.Transactions
+	if len(bodies) != len(headers) {
+		return errors.New("requested body and header mismatch")
```

Comment:
Why is this not a standardized package error?

Reply:
Fixed 👍 

[0e4c69b](https://github.com/0xPolygon/polygon-edge/pull/548/commits/0e4c69b7e4de64ad295640a9a83c65cb65bd28e5)